[SROA] Optimize reloaded values in allocas that escape into readonly nocapture calls.

davemgreen opened the pull request:
Given an alloca that potentially has many uses in big complex code and escapes into a call that is readonly+nocapture, we cannot easily split up the alloca.  There are several optimizations that will attempt to take a value that is stored and a reload, and replace the load with the original stored value.  Instcombine has some simple heuristics, GVN can sometimes do it, as can CSE in limited situations. They all suffer from the same issue with complex code - they start from a load/store and need to prove no-alias for all code between, which in complex cases might be a lot to look through. Especially if the ptr is an alloca with many uses that is over the normal escape capture limits.

The pass that does do well with allocas is SROA, as it has a complete view of all of the uses. This patch adds a case to SROA where it can detect allocas that are passed into calls that are no-capture readonly. It can then optimize the reloaded values inside the alloca slice with the stored value knowing that it is valid no matter the location of the loads/stores from the no-escaping nature of the alloca.

nikic reviewed (commented):
I like the idea. This is kind of similar to load-only promotion in LICM.

Your current implementation handles a single dominating store -- should we use SSAUpdater / LoadStorePromoter instead to handle the general case?

nikic commented on llvm/include/llvm/Analysis/PtrUseVisitor.h:
```diff
@@ -100,6 +113,7 @@ class PtrUseVisitorBase {
   private:
     Instruction *AbortedInfo = nullptr;
     Instruction *EscapedInfo = nullptr;
+    Instruction *EscapedReadOnly = nullptr;
```
I'm concerned that, as implemented, this may affect other users of PtrUseVisitor, which only check isEscaped() and don't know that they need to check isEscapedReadOnly() as well. A read-only escape should probably still report as an escape?

nikic replied:
Looking more closely, I see that the code calling setEscapedReadOnly is part of SROA only rather than the generic implementation, so this is not an issue.

nikic commented on llvm/lib/Transforms/Scalar/SROA.cpp:
```diff
@@ -1390,14 +1394,31 @@ class AllocaSlices::SliceBuilder : public PtrUseVisitor<SliceBuilder> {
+        !CB.hasOperandBundles() && all_of(enumerate(CB.args()), [&](auto V) {
+          return V.value() != *U ||
+                 (Callee->getArg(V.index())->hasNoCaptureAttr() &&
+                  Callee->getArg(V.index())->onlyReadsMemory());
```
Please use the methods on CallBase instead, like `CB.doesNotCapture(Idx)`.

davemgreen commented:
> It looks like this miscompiles stage2 clang in a way that causes the subsequent test-suite build to fail: https://llvm-compile-time-tracker.com/show_error.php?commit=5720d0797414c8413dc797e3db8b1fd8ed324977

Thanks I'll take a look. It looks like I had misunderstood how partitions work with unsplittable slices and how they can overlap.

davemgreen commented:
The issues I found with SSAUpdater was that it loves to turn loads into poison, which is not generally valid for allocas. I've added an option to turn those values in zero instead, let me know if there is a better suggestion.

nikic commented:
> The issues I found with SSAUpdater was that it loves to turn loads into poison, which is not generally valid for allocas. I've added an option to turn those values in zero instead, let me know if there is a better suggestion.

I think the correct way to handle this is to call `AddAvailableValue` for the entry block with an UndefValue.

nikic commented on llvm/lib/Transforms/Scalar/SROA.cpp:
```diff
@@ -1390,14 +1395,30 @@ class AllocaSlices::SliceBuilder : public PtrUseVisitor<SliceBuilder> {
+    // If the operands that are U are NoCapture ReadOnly, then we mark it as
+    // EscapedReadOnly.
+    Function *Callee = CB.getCalledFunction();
+    if (Callee && CB.arg_size() == Callee->arg_size() &&
```
The Callee checks shouldn't be necessary.

nikic commented on llvm/lib/Transforms/Scalar/SROA.cpp:
```diff
@@ -1390,14 +1395,30 @@ class AllocaSlices::SliceBuilder : public PtrUseVisitor<SliceBuilder> {
+    // EscapedReadOnly.
+    Function *Callee = CB.getCalledFunction();
+    if (Callee && CB.arg_size() == Callee->arg_size() &&
+        !CB.hasOperandBundles() && all_of(enumerate(CB.args()), [&](auto V) {
```
The operand bundle checks shouldn't be necessary. doesNotCapture/onlyReadsMemory both support operand bundle indices and also handle operand bundle effects.

nikic commented on llvm/lib/Transforms/Scalar/SROA.cpp:
```diff
@@ -1390,14 +1395,30 @@ class AllocaSlices::SliceBuilder : public PtrUseVisitor<SliceBuilder> {
+    Function *Callee = CB.getCalledFunction();
+    if (Callee && CB.arg_size() == Callee->arg_size() &&
+        !CB.hasOperandBundles() && all_of(enumerate(CB.args()), [&](auto V) {
+          return V.value() != *U || (CB.doesNotCapture(V.index()) &&
```
Why does this iterate over the arguments? The use already knows it's operand number.

davemgreen commented:
> I think the correct way to handle this is to call AddAvailableValue for the entry block with an UndefValue.

Yeah I was hoping that would work. The last commit in this PR has that (but backwards, where it was taken out, a306a3f56b7162587002e91d1d5852f6457630bd). I believe that defines the values at the end of the block, so the value is still converted to poison if it is all in one block. I can maybe add something into LoadAndStorePromoter - the part that understands about values throughout a block - to treat the alloca like a store that defines a value.

nikic commented on llvm/lib/Transforms/Utils/SSAUpdater.cpp:
```diff
@@ -468,6 +472,12 @@ void LoadAndStorePromoter::run(const SmallVectorImpl<Instruction *> &Insts) {
         // Remember that this is the active value in the block.
         StoredValue = SI->getOperand(0);
+      } else if (auto *AI = dyn_cast<AllocaInst>(&I)) {
+        // Check if this a alloca, in which case we treat it as a store of
```
```suggestion
        // Check if this an alloca, in which case we treat it as a store of
```

nikic commented on llvm/lib/Transforms/Scalar/SROA.cpp:
```diff
@@ -5454,6 +5475,79 @@ void SROA::clobberUse(Use &U) {
+  }
+
+  Value *getValueToUseForAlloca(Instruction *I) const override {
+    return Constant::getNullValue(ZeroType);
```
```suggestion
    return UndefValue::get(ZeroType);
```

davemgreen replied:
I thought we were trying to remove uses of undef?

nikic replied:
Generally yes. Uninitialized memory is the exception to this currently.

nikic reviewed (commented):
I see, that's annoying. I think your solution is fine...

nikic commented on llvm/test/Transforms/SROA/readonlynocapture.ll:
```diff
@@ -0,0 +1,309 @@
+  store i32 1, ptr %b
+  call void @callee(ptr %a)
+  %l1 = load i32, ptr %a
+  %l2 = load i32, ptr %b
```
Use both %l1 and %l2 here?

nikic commented on llvm/test/Transforms/SROA/readonlynocapture.ll:
```diff
@@ -0,0 +1,309 @@
+  store i32 1, ptr %b
+  call void @callee(ptr %a)
+  %l1 = load volatile i32, ptr %a
+  %l2 = load volatile i32, ptr %b
```
Missing bailout for volatile load?

Should also test atomic.

davemgreen commented on llvm/lib/Transforms/Scalar/SROA.cpp:
```diff
@@ -43,6 +43,7 @@
 #include "llvm/Analysis/GlobalsModRef.h"
 #include "llvm/Analysis/Loads.h"
 #include "llvm/Analysis/PtrUseVisitor.h"
+#include "llvm/Analysis/ValueTracking.h"
```
Apparently its for `isAssumeLikeIntrinsic`.

davemgreen commented on llvm/lib/Transforms/Scalar/SROA.cpp:
```diff
@@ -5454,6 +5475,86 @@ void SROA::clobberUse(Use &U) {
+        if (auto *LI = dyn_cast<LoadInst>(User)) {
+          Type *UserTy = LI->getType();
+          // LoadAndStorePromoter requires all the types to be the same.
+          if (!LI->isSimple() || (PartitionType && UserTy != PartitionType))
```
I was wondering if it was valid to forward out of a volatile store into a load, but early cse seems happy enough to do it: https://llvm.godbolt.org/z/PzKzGPbE8. I have removed it in the latest patch.

davemgreen committed:
[SROA] Optimize reloaded values in allocas that escape into readonly nocapture calls.

Given an alloca that potentially has many uses in complex code and escapes into
a call that is readonly+nocapture, we cannot easily split up the alloca.  There
are several optimizations that will attempt to take a value that is stored and
a reload, and replace the load with the original stored value.  Instcombine has
some simple heiristics, GVN can sometimes do, as can early CSE in limited
situations. They all suffer from the same issue with complex code - they start
from a load/store and need to prove no-alias for all code between, which in
complex cases might be a loti to look through. Especially if the ptr is an
alloca with many uses that is over the normal escape capture limits.

The pass that does do well with allocas is SROA, as it has a complete view of
the alloca and all of its uses. This patch adds a case to SROA where it can
detect allocas that are passed into calls that are no-capture readonly. It can
then optimize the reloaded values inside the alloca slice with the stored
value knowing that it is valid no matter the location of the loads/stores from
the no-escaping nature of the alloca.

nikic reviewed (approved):
LGTM

davemgreen commented:
Thanks

kstoimenov commented:
Here is the link to the bot: https://lab.llvm.org/buildbot/#/builders/52/builds/4430/steps/13/logs/stdio
